src/Numerical/LinearAlgebra/Determinant.hpp

```diff
@@ -0,0 +1,201 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines function for taking the determinant of a rank-2 tensor
+
+#pragma once
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+namespace detail {
+template <typename Symm, typename Index, typename = void>
+struct DeterminantImpl;
+
+template <typename Symm, typename Index>
+struct DeterminantImpl<Symm, Index, std::enable_if_t<Index::dim == 1>> {
+  template <typename T>
+  static auto apply(const T& tensor) {
+    DataVector det(tensor.template get<0, 0>().size());
+    for (size_t s = 0; s < det.size(); ++s) {
+      det[s] = tensor.template get<0, 0>()[s];
+    }
+    return det;
+  }
+};
+
+template <typename Symm, typename Index>
+struct DeterminantImpl<Symm, Index, std::enable_if_t<Index::dim == 2>> {
+  template <typename T>
+  static auto apply(const T& tensor) {
+    DataVector det(tensor.template get<0, 0>().size());
+    for (size_t s = 0; s < det.size(); s++) {
```

Comment:
Let's use the expression templates everywhere. When I optimize the code I can come back to determinant, if it even matters at all.